### PR TITLE
Query node_meta info using DNS TXT query API

### PR DESCRIPTION
**Pull request by sodre**
This is the node-part of feature request #2709

 - [x] Update documentation
 - [x] Return results for TXT query
 - [x] Return results for ANY query
 - [x] Return resuts for recursive DNS queries
 - [x] Request review by hashicorp (@slackpad)
 - [x] Merge & test with 0.9.1 
 - [x] Write RFC1464 encoder and test cases
 - [ ] Rebase and reorder commits

### CODE REVIEW

File: agent/dns.go [hunk 1]
```diff
@@ -429,39 +429,52 @@ RPC:
...
 	ip := net.ParseIP(addr)
 	var ipv4 net.IP
 	if ip != nil {
 		ipv4 = ip.To4()
 	}
+
 	switch {
 	case ipv4 != nil && (qType == dns.TypeANY || qType == dns.TypeA):
-		return []dns.RR{&dns.A{
+		records = append(records, &dns.A{
```
Comment (sodre) [inline agent/dns.go]: We have to append in order to support ANY queries when the node also has a metadata key-values

File: website/source/docs/agent/dns.html.md [hunk 2]
```diff
@@ -57,8 +57,9 @@ we can instead use `foo.node.consul.` This convention allows for terse
 syntax where appropriate while supporting queries of nodes in remote
 datacenters as necessary.
 
-For a node lookup, the only records returned are A records containing
-the IP address of the node.
+For a node lookup, the only records returned are A and AAAA records
```
Comment (sodre) [inline website/source/docs/agent/dns.html.md]: I added language for 'AAAA' records as well since it is in available in the code. Feel free to remove that in case it is not officially supported yet.

**Comment by magiconair** [comment]
I can have a look at this since I'm in that codepath right now. Please note that we will merge #3353 first.

**Comment by sodre** [comment]
working on solving the conflicts with dns.go

**Comment by sodre** [comment]
@magiconair,
There was as an internal API change/refactoring which removed the `Node` struct out of the `formatNodeRecord` function. 

In order to solve this conflict, I will likely need to change every call to formatNodeRecord to include `node.Meta`.

Please advise if you are okay with that change.

File: agent/dns.go [hunk 3]
```diff
@@ -320,7 +320,7 @@ func (d *DNSServer) nameservers(edns bool) (ns []dns.RR, extra []dns.RR) {
 		ns = append(ns, nsrr)
 
 		// A or AAAA glue record
-		glue := d.formatNodeRecord(ip.String(), name, dns.TypeANY, d.config.NodeTTL, edns)
+		glue := d.formatNodeRecord(nil, ip.String(), name, dns.TypeANY, d.config.NodeTTL, edns)
```
Comment (sodre) [inline agent/dns.go]: Undo `formatNodeRecord` change from commit 76319f751d09a8c77ff9cff8302ce089b17a4ec5

File: agent/dns.go [hunk 4]
```diff
@@ -509,45 +509,58 @@ RPC:
 	n := out.NodeServices.Node
 	edns := req.IsEdns0() != nil
 	addr := d.agent.TranslateAddress(datacenter, n.Address, n.TaggedAddresses)
-	records := d.formatNodeRecord(addr, req.Question[0].Name, qType, d.config.NodeTTL, edns)
+	records := d.formatNodeRecord(out.NodeServices.Node, addr, req.Question[0].Name, qType, d.config.NodeTTL, edns)
```
Comment (sodre) [inline agent/dns.go]: Undo `formatNodeRecord` change from commit 76319f751d09a8c77ff9cff8302ce089b17a4ec5

File: agent/dns.go [hunk 5]
```diff
@@ -509,45 +509,58 @@ RPC:
...
+			txt = append(txt, v)
+		} else {
+			txt = append(txt, k+"="+v)
+		}
+	}
+	return txt
+}
+
+// formatNodeRecord takes a Node and returns an A, AAAA, TXT or CNAME record
+func (d *DNSServer) formatNodeRecord(node *structs.Node, addr, qName string, qType uint16, ttl time.Duration, edns bool) (records []dns.RR) {
```
Comment (sodre) [inline agent/dns.go]: Undo `formatNodeRecord` change from commit 76319f751d09a8c77ff9cff8302ce089b17a4ec5

File: agent/dns.go [hunk 6]
```diff
@@ -575,6 +588,21 @@ func (d *DNSServer) formatNodeRecord(addr, qName string, qType uint16, ttl time.
 			}
 		}
 	}
+
+	if node != nil && len(node.Meta) > 0 && (qType == dns.TypeANY || qType == dns.TypeTXT) {
```
Comment (sodre) [inline agent/dns.go]: Add test in case we are called from line [323](https://github.com/hashicorp/consul/pull/3343/commits/f498c53dc8b1e7aaf06d6bd291428bf07568a879#diff-0f2130d66e011e1af6e42f93492a3536R323).

File: agent/dns.go [hunk 7]
```diff
@@ -896,7 +924,7 @@ func (d *DNSServer) serviceNodeRecords(dc string, nodes structs.CheckServiceNode
 		handled[addr] = struct{}{}
 
 		// Add the node record
-		records := d.formatNodeRecord(addr, qName, qType, ttl, edns)
+		records := d.formatNodeRecord(node.Node, addr, qName, qType, ttl, edns)
```
Comment (sodre) [inline agent/dns.go]: Undo `formatNodeRecord` change from commit 76319f751d09a8c77ff9cff8302ce089b17a4ec5

File: agent/dns.go [hunk 8]
```diff
@@ -940,7 +968,7 @@ func (d *DNSServer) serviceSRVRecords(dc string, nodes structs.CheckServiceNodes
 		}
 
 		// Add the extra record
-		records := d.formatNodeRecord(addr, srvRec.Target, dns.TypeANY, ttl, edns)
+		records := d.formatNodeRecord(node.Node, addr, srvRec.Target, dns.TypeANY, ttl, edns)
```
Comment (sodre) [inline agent/dns.go]: Undo `formatNodeRecord` change from commit 76319f751d09a8c77ff9cff8302ce089b17a4ec5

**Comment by sodre** [comment]
This is ready for review again. 
The travis-ci build that failed looked unrelated to my changes.

File: agent/dns.go [hunk 9]
```diff
@@ -509,45 +509,58 @@ RPC:
...
-	records := d.formatNodeRecord(addr, req.Question[0].Name, qType, d.config.NodeTTL, edns)
+	records := d.formatNodeRecord(out.NodeServices.Node, addr, req.Question[0].Name, qType, d.config.NodeTTL, edns)
 	if records != nil {
 		resp.Answer = append(resp.Answer, records...)
 	}
 }
 
-// formatNodeRecord takes a Node and returns an A, AAAA, or CNAME record
-func (d *DNSServer) formatNodeRecord(addr, qName string, qType uint16, ttl time.Duration, edns bool) (records []dns.RR) {
+// formatTxtRecords takes a kv-map and returns it as "[k=]v" for non-empty k not starting with rfc1035-
```
Comment (magiconair) [inline agent/dns.go]: * I'd suggest that you describe what the function does and not how it does that, e.g.

> formatTxtRecords encodes the map values according to RFC 1464.

* `formatTxtRecords` isn't a good name since it doesn't return records (`[]*dns.RR`) but only the values of `TXT` records. Maybe name it `encodeKVasRFC1464` or something similar. I didn't think this all the way through but you should get the drift.

* The comment does not match the code since `k` can be empty and you'd end up with `=v`.

* RFC 1464 says that key names with special characters must be encoded:

https://www.ietf.org/rfc/rfc1464.txt
>    Attribute Names
> 
>    Any printable ASCII character is permitted for the attribute name.
>    If an equals sign is embedded in the attribute name, it must be
>    quoted with a preceding grave accent (or backquote: ``"`"``).  A
>    backquote must also be quoted with an additional ``"`"``.

also further down:

>    Leading and trailing whitespace (spaces and tabs) in the attribute
>    name are ignored unless they are quoted (with a "`").  For example,
>    "abc" matches " abc<tab>" but does not match ``"` abc"``.
> 
>    Note that most DNS server implementations require a backslash (\) or
>    double quote (") in a text string to be quoted with a preceding
>    backslash.  Accent grave (``"`"``) was chosen as a quoting character in
>    this syntax to avoid confusion with "\" (and remove the need for
>    confusing strings that include sequences like "\\\\").
> 

and

>    Attribute Values
> 
>    All printable ASCII characters are permitted in the attribute value.
>    No characters need to be quoted with a "`".  In other words, the
>    first unquoted equals sign in the TXT record is the name/value
>    delimiter.  All subsequent characters are part of the value.
> 
>    Once again, note that in most implementations the backslash character
>    is an active quoting character (and must, itself, be quoted).
> 

* I'd prefer a `switch` statement over `if/else`
* Please add a table driven test for this function especially the edge cases

* When looking at #2709 I've noticed the use of `datacenter` in the example. Would that be the consul datacenter value and if yes how does this get in here? We may want to think about some expansion language like `${DC}` or similar to fill in values from the other consul node data but that may be overkill.

File: agent/dns.go (hunk 6 above)
Comment (magiconair) [inline agent/dns.go]: `len(node.Meta) > 0` is redundant

File: agent/dns_test.go [hunk 10]
```diff
@@ -351,6 +364,88 @@ func TestDNS_NodeLookup_CNAME(t *testing.T) {
...
+	m.SetQuestion("google.node.consul.", dns.TypeTXT)
+
+	c := new(dns.Client)
+	addr, _ := a.Config.ClientListener("", a.Config.Ports.DNS)
+	in, _, err := c.Exchange(m, addr.String())
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Should have the 1 TXT record reply
```
Comment (magiconair) [inline agent/dns_test.go]: Pls use `verify.Values` to compare full responses, if possible.

File: agent/dns_test.go [hunk 11]
```diff
@@ -351,6 +364,88 @@ func TestDNS_NodeLookup_CNAME(t *testing.T) {
...
+
+	c := new(dns.Client)
+	addr, _ := a.Config.ClientListener("", a.Config.Ports.DNS)
+	in, _, err := c.Exchange(m, addr.String())
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Should have 1 A and 1 TXT record
+	if len(in.Answer) != 2 {
```
Comment (magiconair) [inline agent/dns_test.go]: Pls `verify.Values`, if possible.

File: agent/dns_test.go [hunk 12]
```diff
@@ -334,8 +347,8 @@ func TestDNS_NodeLookup_CNAME(t *testing.T) {
 		t.Fatalf("err: %v", err)
 	}
 
-	// Should have the service record, CNAME record + A record
-	if len(in.Answer) != 3 {
+	// Should have the service record, CNAME record + A + TXT record
```
Comment (magiconair) [inline agent/dns_test.go]: > Should have the service, CNAME, A, and TXT records

File: agent/dns.go (hunk 3 above)
Comment (magiconair) [inline agent/dns.go]: This function signature is starting to get a bit out of hand but that's a story for another day.

File: agent/dns.go [hunk 13]
```diff
@@ -464,9 +464,9 @@ INVALID:
 
 // nodeLookup is used to handle a node query
 func (d *DNSServer) nodeLookup(network, datacenter, node string, req, resp *dns.Msg) {
-	// Only handle ANY, A and AAAA type requests
+	// Only handle ANY, A, AAAA and TXT type requests
```
Comment (magiconair) [inline agent/dns.go]: Oxford comma?

File: agent/dns.go (hunk 3 above)
Comment (preetapan) [inline agent/dns.go]: I removed node from there in that commit because it was an unused param. why are we adding it back?

Comment (magiconair) [inline agent/dns.go]: Because he needs it to get to the meta data of the node. We might also just pass the map in.

File: agent/dns.go (hunk 9 above)
Comment (sodre) [inline agent/dns.go]: This PR only closes the TXT lookup side for Nodes. I am open to suggestions on how to solve it at the datacenter level. 

My initial thought was to use the KV system, but that looked like too much of a hack.

Are there any plans to make data-center level metadata values available?

Comment (magiconair) [inline agent/dns.go]: I was referring to a value containing the configured datacenter of consul. Lets consider this when it is requested and get this one done first.

**Comment by sodre** [comment]
I added the tests from RFC1464. 
Because of the [restrictions on node-meta](https://www.consul.io/docs/agent/options.html#_node_meta) Some of the tests cases from RFC1464 would never get executed in the live system, e.g. '=', or empty keys.

**Comment by sodre** [comment]
@magiconair 
Do you need any additional changes? If not I would like to reorder some commits and squash a few things before the final merge.

Thanks!

**Comment by magiconair** [comment]
@sodre I didn't get a chance to review your latest changes yet. I'll ping you when I'm done. If you want and there are no changes then I can do the re-order squash for you if you let me know what you want to do.

**Comment by sodre** [comment]
@magiconair 
Can we get a milestone for you to review this piece?  I want to avoid fixing the conflict now, and another one popping up because the merge is taking too long.

**Comment by magiconair** [comment]
Hi @sodre, sorry for the delay. I was on vacation for a couple of weeks and have two large things I'm working on right now. That's why I'm avoiding to context switch ATM. I'll try to review it this week.

**Comment by magiconair** [comment]
I've rebased the code and fixed the tests but I'm not getting the result I thought:

```
$ consul agent -dev -node-meta "baz:bang" -node-meta 'blurb' -node foo

$ dig @localhost -p 8600 foo.node.consul. any

; <<>> DiG 9.8.3-P1 <<>> @localhost -p 8600 foo.node.consul. any
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 32778
;; flags: qr aa rd; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;foo.node.consul.		IN	ANY

;; ANSWER SECTION:
foo.node.consul.	0	IN	A	127.0.0.1
foo.node.consul.	0	IN	TXT	"consul-network-segment="
foo.node.consul.	0	IN	TXT	"blurb="
foo.node.consul.	0	IN	TXT	"baz=bang"
```

I thought meta keys without a value shouldn't have a trailing `=` sign.

**Comment by magiconair** [comment]
I've pushed my changes to the https://github.com/hashicorp/consul/tree/zeroae-f-node-dns-txt-record-magiconair branch. I'll see whether I can fix this later today.

**Comment by magiconair** [comment]
Just re-read https://tools.ietf.org/html/rfc1464 so I think we're good.

```
   Examples

   <sp> indicates a space character.

   Attribute    Attribute       Internal Form           External Form
   Name         Value           (server to resolver)    (TXT record)

   color        blue            color=blue              "color=blue"
   equation     a=4             equation=a=4            "equation=a=4"
   a=a          true            a`=a=true               "a`=a=true"
   a\=a false           a\`=a=false             "a\\`=a=false"
   =            \=              `==\=                   "`==\\="

   string       "Cat"           string="Cat"            "string=\"Cat\""
   string2      `abc`           string2=``abc``         "string2=``abc``"


   novalue                      novalue=                "novalue="   <-- this


   a b          c d             a b=c d                 "a b=c d"
   abc<sp>      123<sp>         abc` =123<sp>           "abc` =123 "

```
